server/app/middlewares/websocket_backend.rb

```diff
@@ -67,29 +75,104 @@ def call(env)
     end
   end
 
-  ##
-  # On websocket connection open
-  #
-  # @param [Faye::WebSocket] ws
-  # @param [Rack::Request] req
-  def on_open(ws, req)
+  # @param node_id [String] request header Kontena-Node-Id
+  # @param grid_token [String] request header Kontena-Grid-Token
+  # @param init_attrs [Hash] initialize attributes on new node
+  # @raise [CloseError]
+  # @return [HostNode] with node_id set
+  def find_node_by_grid_token(node_id, grid_token, init_attrs)
     # check grid
-    grid = Grid.find_by(token: req.env['HTTP_KONTENA_GRID_TOKEN'].to_s)
-    unless grid
-      logger.error 'invalid grid token, closing connection'
-      ws.close(4001, "Invalid grid token")
-      return
+    grid = Grid.find_by(token: grid_token.to_s)
+
+    raise CloseError.new(4001), "Invalid grid token" unless grid
+
+    node = grid.host_nodes.find_by(node_id: node_id)
+
+    if !node
+      node = grid.host_nodes.create!(node_id: node_id, **init_attrs)
+
+      logger.info "new node #{node} connected using grid token"
+
+    elsif node.token
+      raise CloseError.new(4005), "Invalid grid token, node was created using a node token"
+
+    else
+      logger.debug "node #{node} connected using grid token"
     end
 
-    # check host node
+    return node
+  end
+
+  # @param node_id [String] request header Kontena-Node-Id
+  # @param node_token [String] request header Kontena-Node-Token
+  # @param init_attrs [Hash] initialize attributes on new node
+  # @raise [CloseError]
+  # @return [HostNode] with node_id set
+  def find_node_by_node_token(node_id, node_token, init_attrs)
+    node = HostNode.find_by(token: node_token.to_s)
+
+    raise CloseError.new(4002), "Invalid node token" unless node
+
+    node_by_id = HostNode.find_by(node_id: node_id)
+
+    if !node_by_id
+      # atomically initialize the node_id
+      initializing_node = HostNode.where(:id => node.id, :node_id => nil)
+        .find_one_and_update(:$set => {node_id: node_id, **init_attrs})
+
+      if initializing_node
+        logger.info "new node #{node} connected using node token with node_id #{node_id}"
+
+        node.reload
+
+      else
+        logger.warn "new node #{node} connected using node token with node_id #{node_id}, but the node token was already used by #{node} with node ID #{node.node_id}"
+
+        raise CloseError.new(4003), "Invalid node token, already used by a different node"
+      end
+
+    elsif node.id == node_by_id.id
+      logger.debug "node #{node} connected using node token with node_id #{node_id}"
+
+    else
+      logger.warn "node #{node} connected using node token with node_id #{node_id}, but that node ID already exists for #{node_by_id}"
+
+      raise CloseError.new(4006), "Invalid node ID, already used by a different node"
```

Comment:
This happens if the same node, or two cloned nodes that have the same Docker ID, both connect with unique `kontena node create` -> node tokens: one of them connects first, and claims the node ID but setting it on the node identified by the token.

The second node to connect will find its own node by the token, and then find the other node by the node ID. The two will have different BSON ID:s, and this will fail.

There's a brief race condition window between two such nodes connecting at the same time, and both of them lookup their own node ID before doing the `find_one_and_update`. The second one go down in flames with a `Mongo::Error::OperationFailure`, like in the grid token case.